### DISCUSSION TITLE
posts/dca-expose-port

### DISCUSSION THREAD

**Discussion by giscus** [allengaodev/allengaodev.github.io#214] [Announcements]
# posts/dca-expose-port

Docker Certified Associate(DCA)認證相關資源、Docker Port Binding、Docker EXPOSE、開放端口

https://blog.allengaodev.com/posts/dca-expose-port

**Comment by fbukevin** [comment]
從前面有一個的例子來看：
```shell
[node1] (local) root@192.168.0.8 ~
$ docker ps -a
CONTAINER ID   IMAGE     COMMAND                  CREATED         STATUS         PORTS                          NAMES
b538e3d3eec1   nginx     "/docker-entrypoint.…"   7 seconds ago   Up 6 seconds   80/tcp, 0.0.0.0:8080->81/tcp   clever_einstein
```

Dockerfille 中應該只有 EXPOSE 80 沒有 81
因此如果 Dockerfile 中沒有 EXPOSE，只要 docker run 有用 -p binding，就會自動 expose 那個 port
這樣理解對嗎？(我會來做個實驗測試)

**Reply by allengaodev** [reply]
因為 Nginx 預設的 nginx.conf 就是 listen 80 port，所以他們在編寫 Dockerfile 的時候有在內部 EXPOSE 80 Port，這樣 Nginx 容器的使用者就會知道這個容器內的應用程式要使用 80 Port，因為容器預設是使用 `bridge` 模式，所以你的電腦網路跟容器的網路的網路是處於隔離狀態，因此就需要透過綁定的方式來將兩個網路環境對應在一起。

`docker run -dit --name nginx28 -p 8080:81 nginx` 
這裡的 `-p 8080:81` 是指使用者想要綁定主機的 8080 Port 到容器的 81 Port 的意思，並沒有 Expose 81 Port。

想要測試的話可以建立一個 `-p 8080:81` 的 nginx 容器，使用 exec 進入到容器內部使用 netstat 指令檢查目前 Listen 的 Port，就會發現只有 80 Port。
